src/transform/transform.ts

```diff
@@ -938,3 +939,96 @@ function parseComments(comment: string): string {
     tables: true,
   });
 }
+
+
+// This function try to label all unreferenced types before doing transform.
+async function labelUnreferencedTypes(session: Session<CodeModel>) {
+  const model = session.model;
+
+  const isRemoveUnreferencedTypes = await session.getValue('remove-unreferenced-types', false);
+
+  if (!isRemoveUnreferencedTypes) return;
+
+  const referencedTypes = new Set<Schema>();
+
+  iterateOperations(model, referencedTypes);
+
+  labelOmitTypes(model.schemas.choices, referencedTypes);
+  labelOmitTypes(model.schemas.sealedChoices, referencedTypes);
+  labelOmitTypes(model.schemas.objects, referencedTypes);
+}
+
+// This function help to label omit types according to the referencedTypes set.
+function labelOmitTypes<Type extends Schema>(modelSchemas: Array<Type> | undefined, referencedSet: Set<Type>) {
+  if (modelSchemas){
+    for (const schema of modelSchemas){
+      if (!referencedSet.has(schema)){
+        schema.language.go!.omitType = true;
+      }
+    }
+  }
+}
+
+// This function iterate all the operations in all operation groups and try to find all the referenced types (including choices, sealedChoices and objects).
+// For each operation, we will aggregate all the parameters' type of the operation and put them into the referencedTypes set.
+// Also, all responses body types and header types will be put into the referencedTypes set.
+// Some special cases: exceptions response types, x-ms-odata types.
+function iterateOperations(model: CodeModel, referencedTypes: Set<Schema>) {
+  for (const group of values(model.operationGroups)) {
+    for (const op of values(group.operations)) {
+      for (const param of values(aggregateParameters(op))) {
+        dfsSchema(param.schema, referencedTypes);
+      }
+      // We do not use error responses' definition to unmarshal response. So exceptions will be ignored.
+      // const responses = values(op.responses).concat(values(op.exceptions));
+      const responses = values(op.responses);
+      for (const resp of values(responses)) {
+        if (isSchemaResponse(resp)) {
+          dfsSchema(resp.schema, referencedTypes);
+        }
+        const httpResponse = <HttpResponse>resp.protocol.http;
+        for (const header of values(httpResponse.headers)) {
+          dfsSchema(header.schema, referencedTypes);
+        }
+      }
+      // Such odata definition is not used directly in operation. But it seems the model is a detailed definition for some string param. Reserve for now.
+      if (op.extensions?.['x-ms-odata']) {
+        const schemaParts = op.extensions['x-ms-odata'].split('/');
+        const schema = values(model.schemas.objects).where((o) => o.language.default.name === schemaParts[schemaParts.length - 1]).first();
+        if (schema) {
+          dfsSchema(schema, referencedTypes);
+        }
+      }
```

Comment:
if we have to parse the odata extension in this way by splitting the path, does this mean the modeler v4 does not do anything to this extension?

Reply:
It seems yes.